mesos-client/src/main/scala/com/mesosphere/mesos/client/CredentialsProvider.scala

```diff
@@ -0,0 +1,78 @@
+package com.mesosphere.mesos.client
+
+import java.net.URL
+import java.time.Instant
+
+import akka.NotUsed
+import akka.actor.ActorSystem
+import akka.http.scaladsl.Http
+import akka.http.scaladsl.model._
+import akka.http.scaladsl.model.headers.{BasicHttpCredentials, GenericHttpCredentials, HttpCredentials}
+import akka.http.scaladsl.unmarshalling.Unmarshal
+import akka.stream.ActorMaterializer
+import akka.stream.scaladsl.Source
+import org.json4s.native.JsonMethods
+import pdi.jwt.JwtJson4s
+import pdi.jwt.JwtAlgorithm.RS256
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.duration._
+
+trait CredentialsProvider {
+
+  def tokens(): Source[HttpCredentials, NotUsed]
+}
+
+case class JwtProvider(uid: String, privateKey: String, root: URL)(implicit system: ActorSystem, materializer: ActorMaterializer, context: ExecutionContext) extends CredentialsProvider {
+  import org.json4s._
+  import org.json4s.JsonDSL._
+  import JsonMethods.{parse, render, compact}
+
+  private def expireIn(duration: Duration): Long = Instant.now.getEpochSecond + duration.toSeconds
+  private val claim = JObject(("uid", uid), ("exp", expireIn(60.minutes))) // TODO: configure token expiration.
+
+  val acsTokenRequest: HttpRequest = {
+    val token = JwtJson4s.encode(claim, privateKey, RS256)
+    val data: String = compact(render(JObject(("uid", uid), ("token", token))))
+
+    HttpRequest(
+      method = HttpMethods.POST,
+      uri = Uri(s"${root.getPath}/acs/api/v1/auth/login"),
+      entity = HttpEntity(ContentTypes.`application/json`, data)
+    )
+  }
+
+  /**
+    * Provides a [[Source]] of DC/OS acs tokens wrapped in [[HttpCredentials]]. The token is refreshed on each pull.
+    *
+    * @return A new source of [[HttpCredentials]]
+    */
+  def tokens(): Source[HttpCredentials, NotUsed] = {
+    val port = if(root.getPort == -1) root.getDefaultPort else root.getPort
+
+    Source.repeat(acsTokenRequest)
+      .via(Http().outgoingConnectionHttps(root.getHost, port))
+      .mapAsync(1) { response =>
```

Comment:
After thinking about this, I don't think Akka streams should be used for this.

1. It's not as lazy as you think. MapAsync adds an implicit buffer which will cause multiple http requests to be issued ahead of time to fill the buffer before any elements are requested; therefore, by the time they are requested, they are stale.
2. This concern is just as easily handled by a method which can be called, on demand, and return a future. Making it an Akka stream source provides virtually no benefit over this.

Reply:
Yeah, I thought there would be an easy way to first pull from the tokens stream and then just consume Mesos calls. I'm working on a simple graph stage today and hope that it''l work out.